crates/valence_advancement/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+pub mod event;
+
+use std::borrow::Cow;
+use std::io::Write;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use anyhow::Context;
+use bevy_app::{CoreSet, Plugin};
+use bevy_ecs::prelude::{Bundle, Component, Entity};
+use bevy_ecs::query::{Added, Changed, Or, With};
+use bevy_ecs::schedule::{IntoSystemConfig, IntoSystemSetConfig, SystemSet};
+use bevy_ecs::system::{Commands, Query, SystemParam};
+use bevy_ecs::world::Ref;
+#[doc(hidden)]
+pub use bevy_hierarchy;
+use bevy_hierarchy::{Children, Parent};
+use event::{handle_advancement_tab_change, AdvancementTabChange};
+use parking_lot::Mutex;
+use rustc_hash::FxHashMap;
+use valence_client::{Client, FlushPacketsSet, SpawnClientsSet};
+use valence_core::__private::VarInt;
+use valence_core::ident::Ident;
+use valence_core::item::ItemStack;
+use valence_core::packet::encode::WritePacket;
+use valence_core::packet::s2c::play::{AdvancementUpdateS2c, SelectAdvancementTabS2c};
+use valence_core::packet::{Encode, Packet};
+use valence_core::text::Text;
+
+pub struct AdvancementPlugin;
+
+#[derive(SystemSet, Clone, Copy, Eq, PartialEq, Hash, Debug)]
+pub struct WriteAdvancementPacketToClientsSet;
+
+impl Plugin for AdvancementPlugin {
+    fn build(&self, app: &mut bevy_app::App) {
+        app.configure_set(
+            WriteAdvancementPacketToClientsSet
+                .in_base_set(CoreSet::PostUpdate)
+                .before(FlushPacketsSet),
+        )
+        .add_event::<AdvancementTabChange>()
+        .add_system(
+            init_clients
+                .after(SpawnClientsSet)
+                .in_base_set(CoreSet::PreUpdate),
+        )
+        .add_system(handle_advancement_tab_change.in_base_set(CoreSet::PreUpdate))
+        .add_system(send_advancement_update_packet.in_set(WriteAdvancementPacketToClientsSet));
+    }
+}
+
+#[derive(Bundle)]
+pub struct AdvancementBundle {
+    pub advancement: Advancement,
+    cached_bytes: AdvancementCachedBytes,
+}
+
+impl AdvancementBundle {
+    pub fn new(ident: Ident<Cow<'static, str>>) -> Self {
+        Self {
+            advancement: Advancement::new(ident),
+            cached_bytes: AdvancementCachedBytes(Mutex::new(vec![])),
+        }
+    }
+}
+
+fn init_clients(mut commands: Commands, query: Query<Entity, Added<Client>>) {
```

Comment:
I would prefer if this were named something more descriptive like `add_advancement_component_to_new_clients` to help when reading the output from `bevy_mod_debugdump`.